File: firmware/protocol.cpp
#include "protocol.h"
#include "config.h"
#include <string.h>
#include <stdlib.h>
#include <stdio.h>
#include <math.h>

// ============================================================================
// XOR checksum calculation
// ============================================================================
static uint8_t calc_checksum(const char *data, size_t len)
{
    uint8_t cs = 0;
    for (size_t i = 0; i < len; i++) {
        cs ^= static_cast<uint8_t>(data[i]);
    }
    return cs;
}

// ============================================================================
// Frame parser (format: $TYPE,PAYLOAD*XX)
// ============================================================================
bool parse_frame(const char *line, size_t len, ParsedFrame &out)
{
    out.type = FrameType::UNKNOWN;

    if (len < 6 || line[0] != '$') return false;

    // Find '*' checksum separator
    const char *star = static_cast<const char *>(memchr(line, '*', len));
    if (!star) return false;

    // Data section: between $ and *
    const char *data_start = line + 1;
    size_t data_len = star - data_start;
    if (data_len < 3) return false;

    // Verify checksum
    if (star + 3 > line + len) return false;
    char cs_buf[3] = {star[1], star[2], '\0'};
    uint8_t rx_cs = static_cast<uint8_t>(strtol(cs_buf, nullptr, 16));
    if (rx_cs != calc_checksum(data_start, data_len)) return false;

    // ---- TEMP_CPU / TEMP_GPU ----
    if (data_len >= 6 && (strncmp(data_start, "CPU,", 4) == 0 ||
                          strncmp(data_start, "GPU,", 4) == 0))
    {
        bool is_cpu = (data_start[0] == 'C');
        out.type = is_cpu ? FrameType::TEMP_CPU : FrameType::TEMP_GPU;
        float temp = strtof(data_start + 4, nullptr);
        // P1-3: reject NaN, Inf, and out-of-range temperatures
        if (isnan(temp) || isinf(temp)) return false;
        if (temp < -50.0f || temp > 150.0f) return false;
        out.temp.temperature = temp;
        out.temp.valid = true;
        return true;
    }

    // ---- STATUS_QUERY: $STA,?*XX ----
    if (data_len == 5 && strncmp(data_start, "STA,", 4) == 0) {
        out.type = FrameType::STATUS_QUERY;
        return true;
    }

    // ---- FCURVE_QUERY: $FCQ,?*XX ----
    if (data_len == 5 && strncmp(data_start, "FCQ,", 4) == 0) {
        out.type = FrameType::FCURVE_QUERY;
        return true;
    }

    // ---- FCURVE_SET ----
    if (data_len >= 6 && strncmp(data_start, "FCV,", 4) == 0) {
        const char *p = data_start + 4;

        // Parse point count N
        uint8_t count = static_cast<uint8_t>(strtol(p, const_cast<char **>(&p), 10));
        if (count < 2 || count > 10) return false;
        if (*p != ',') return false;

        out.fcurve.count = count;
        for (uint8_t i = 0; i < count; i++) {
            p++; // skip comma
            float t = strtof(p, const_cast<char **>(&p));
            if (*p != ',') return false;
            p++; // skip comma
            uint8_t d = static_cast<uint8_t>(strtol(p, const_cast<char **>(&p), 10));
            out.fcurve.points[i].temperature = t;
            out.fcurve.points[i].duty_percent = d;
        }

        out.type = FrameType::FCURVE_SET;
        return true;
    }

    // ---- MODE_SET: $MOD,<1-4>*XX ----
    if (data_len >= 5 && strncmp(data_start, "MOD,", 4) == 0) {
        long mode = strtol(data_start + 4, nullptr, 10);
        if (mode < 1 || mode > 4) return false;
        out.ctrl.mode = static_cast<uint8_t>(mode);
        out.type = FrameType::MODE_SET;
        return true;
    }

    // ---- FREQ_SET: $FRQ,<hz>*XX ----
    if (data_len >= 6 && strncmp(data_start, "FRQ,", 4) == 0) {
        long freq = strtol(data_start + 4, nullptr, 10);
        if (freq < PWM_FREQ_MIN || freq > PWM_FREQ_MAX) return false;
        out.ctrl.freq = static_cast<int>(freq);
        out.type = FrameType::FREQ_SET;
        return true;
    }

    // ---- DUTY_SET: $DUT,<0-100>*XX ----
    if (data_len >= 5 && strncmp(data_start, "DUT,", 4) == 0) {
        long duty = strtol(data_start + 4, nullptr, 10);
        if (duty < MIN_SAFE_DUTY_PERCENT || duty > 100) return false;  // P0-4: enforce minimum safe duty
        out.ctrl.duty = static_cast<uint8_t>(duty);
        out.type = FrameType::DUTY_SET;
        return true;
    }

    // ---- SAFETY_RESET: $SAF*XX ----
    if (data_len == 3 && strncmp(data_start, "SAF", 3) == 0) {
        out.type = FrameType::SAFETY_RESET;
        return true;
    }

    return false;
}

// ============================================================================
// Frame finalization: append *XX checksum and \n
// ============================================================================

static size_t finalize_frame(char *buf, size_t buf_size, size_t data_len)
{
    // data_len is length of $TYPE,PAYLOAD in buf
    if (data_len + 4 > buf_size) return 0; // need room for *XX\n\0

    const char *data_start = buf + 1; // skip leading $
    uint8_t cs = calc_checksum(data_start, data_len - 1); // -1 to exclude $ from checksum

    // Append *XX\n to buf at position data_len
    size_t total = data_len + snprintf(buf + data_len, buf_size - data_len, "*%02X\n", cs);
    return total;
}

size_t build_status_response(char *buf, size_t buf_size,
                             int mode, uint8_t duty_pct, int freq_hz,
                             float cpu_temp, float gpu_temp,
                             bool cpu_valid, bool gpu_valid)
{
    int data_len = snprintf(buf, buf_size, "$STP,%d,%d,%d,%.1f,%.1f,%d,%d",
                            mode, duty_pct, freq_hz,
                            cpu_temp, gpu_temp,
                            cpu_valid ? 1 : 0, gpu_valid ? 1 : 0);
    if (data_len < 0 || static_cast<size_t>(data_len) >= buf_size) return 0;
    return finalize_frame(buf, buf_size, static_cast<size_t>(data_len));
}

size_t build_fcurve_response(char *buf, size_t buf_size,
                             const FanCurvePoint *points, uint8_t count)
{
    size_t pos = snprintf(buf, buf_size, "$FCP,%d", count);
    if (pos >= buf_size) return 0;

    for (uint8_t i = 0; i < count; i++) {
        int added = snprintf(buf + pos, buf_size - pos, ",%.1f,%d",
                             points[i].temperature, points[i].duty_percent);
        if (added < 0 || static_cast<size_t>(added) >= buf_size - pos) return 0;
        pos += added;
    }

    return finalize_frame(buf, buf_size, pos);
}

size_t build_ack(char *buf, size_t buf_size)
{
    size_t data_len = snprintf(buf, buf_size, "$ACK");
    if (data_len >= buf_size) return 0;
    return finalize_frame(buf, buf_size, data_len);
}

size_t build_nack(char *buf, size_t buf_size, uint8_t error_code)
{
    int data_len = snprintf(buf, buf_size, "$NAK,%02d", error_code);
    if (data_len < 0 || static_cast<size_t>(data_len) >= buf_size) return 0;
    return finalize_frame(buf, buf_size, static_cast<size_t>(data_len));
}

File: CPUwenduhuoqu/Communication/Protocol.cs
using System;
using System.Globalization;
using System.Text;

namespace CPUwenduhuoqu.Communication
{
    public enum FrameType
    {
        TempCpu,
        TempGpu,
        StatusQuery,
        StatusResponse,
        FcurveSet,
        FcurveQuery,
        FcurveResponse,
        Ack,
        Nack,
        SafetyReset,  // P0-6: remote safety override reset
        Unknown
    }

    public struct StatusData
    {
        public int Mode;
        public int DutyPercent;
        public int FreqHz;
        public float CpuTemp;
        public float GpuTemp;
        public bool CpuValid;
        public bool GpuValid;
    }

    public static class Protocol
    {
        private const string FrameStart = "$";
        private const string ChecksumSep = "*";
        private const string FrameEnd = "\n";

        public static byte ComputeChecksum(string data)
        {
            byte cs = 0;
            foreach (char c in data)
                cs ^= (byte)c;
            return cs;
        }

        private static string FinalizeFrame(string data)
        {
            byte cs = ComputeChecksum(data);
            return FrameStart + data + ChecksumSep + cs.ToString("X2") + FrameEnd;
        }

        // ---- 帧构建 ----

        public static string BuildTempFrame(bool isCpu, float temp)
        {
            string data = (isCpu ? "CPU" : "GPU") + "," + temp.ToString("F1", CultureInfo.InvariantCulture);
            return FinalizeFrame(data);
        }

        public static string BuildStatusQuery()
        {
            return FinalizeFrame("STA,?");
        }

        public static string BuildFcurveQuery()
        {
            return FinalizeFrame("FCQ,?");
        }

        public static string BuildFcurveSet(FanCurvePoint[] points)
        {
            if (points == null || points.Length < 2 || points.Length > 10)
                throw new ArgumentException("风扇曲线必须包含 2-10 个点");

            var sb = new StringBuilder("FCV,");
            sb.Append(points.Length);
            foreach (var p in points)
            {
                sb.Append(',');
                sb.Append(p.Temperature.ToString("F1", CultureInfo.InvariantCulture));
                sb.Append(',');
                sb.Append(p.DutyPercent);
            }
            return FinalizeFrame(sb.ToString());
        }

        public static string BuildModeSet(int mode)
        {
            if (mode < 1 || mode > 4)
                throw new ArgumentException("Mode must be 1-4");
            return FinalizeFrame($"MOD,{mode}");
        }

        public static string BuildFreqSet(int freqHz)
        {
            if (freqHz < 1000 || freqHz > 40000)
                throw new ArgumentException("Frequency must be 1000-40000 Hz");
            return FinalizeFrame($"FRQ,{freqHz}");
        }

        public static string BuildDutySet(int dutyPercent)
        {
            if (dutyPercent < 20 || dutyPercent > 100)  // P0-4: enforce minimum safe duty
                throw new ArgumentException("Duty must be 20-100%");
            return FinalizeFrame($"DUT,{dutyPercent}");
        }

        public static string BuildSafetyReset()  // P0-6: remote safety override reset
        {
            return FinalizeFrame("SAF");
        }

        // ---- 帧解析 ----

        public static FrameType IdentifyFrame(string frame)
        {
            if (string.IsNullOrEmpty(frame) || !frame.StartsWith(FrameStart))
                return FrameType.Unknown;

            int starIdx = frame.IndexOf(ChecksumSep);
            if (starIdx < 0 || starIdx + 3 > frame.Length)
                return FrameType.Unknown;

            // 去除尾部 \r\n
            string cleanFrame = frame.TrimEnd('\r', '\n');

            string dataPart = cleanFrame.Substring(1, starIdx - 1);
            string csStr = cleanFrame.Substring(starIdx + 1, 2);
            byte expectedCs = ComputeChecksum(dataPart);
            byte actualCs;
            if (!byte.TryParse(csStr, NumberStyles.HexNumber, CultureInfo.InvariantCulture, out actualCs))
                return FrameType.Unknown;
            if (expectedCs != actualCs)
                return FrameType.Unknown;

            if (dataPart.StartsWith("CPU,")) return FrameType.TempCpu;
            if (dataPart.StartsWith("GPU,")) return FrameType.TempGpu;
            if (dataPart.StartsWith("STA,")) return FrameType.StatusQuery;
            if (dataPart.StartsWith("STP,")) return FrameType.StatusResponse;
            if (dataPart.StartsWith("FCV,")) return FrameType.FcurveSet;
            if (dataPart.StartsWith("FCQ,")) return FrameType.FcurveQuery;
            if (dataPart.StartsWith("FCP,")) return FrameType.FcurveResponse;
            if (dataPart == "ACK") return FrameType.Ack;
            if (dataPart.StartsWith("NAK,")) return FrameType.Nack;
            if (dataPart == "SAF") return FrameType.SafetyReset;

            return FrameType.Unknown;
        }

        public static bool TryParseStatusResponse(string frame, out StatusData status)
        {
            status = new StatusData();

            if (IdentifyFrame(frame) != FrameType.StatusResponse)
                return false;

            int starIdx = frame.IndexOf(ChecksumSep);
            string dataPart = frame.Substring(1, starIdx - 1);
            // STP,M,D,F,CT,GT,CV,GV
            string[] parts = dataPart.Split(',');
            if (parts.Length != 8) return false;

            try
            {
                status.Mode = int.Parse(parts[1], CultureInfo.InvariantCulture);
                status.DutyPercent = int.Parse(parts[2], CultureInfo.InvariantCulture);
                status.FreqHz = int.Parse(parts[3], CultureInfo.InvariantCulture);
                status.CpuTemp = float.Parse(parts[4], CultureInfo.InvariantCulture);
                status.GpuTemp = float.Parse(parts[5], CultureInfo.InvariantCulture);
                status.CpuValid = parts[6] == "1";
                status.GpuValid = parts[7] == "1";
                return true;
            }
            catch
            {
                return false;
            }
        }

        public static bool TryParseFcurveResponse(string frame, out FanCurvePoint[] points)
        {
            points = null;

            if (IdentifyFrame(frame) != FrameType.FcurveResponse)
                return false;

            int starIdx = frame.IndexOf(ChecksumSep);
            string dataPart = frame.Substring(1, starIdx - 1);
            // FCP,N,t1,d1,...,tN,dN
            string[] parts = dataPart.Split(',');
            if (parts.Length < 3) return false;

            int count;
            if (!int.TryParse(parts[1], out count) || count < 2 || count > 10)
                return false;
            if (parts.Length != 2 + count * 2) return false;

            points = new FanCurvePoint[count];
            for (int i = 0; i < count; i++)
            {
                float temp;
                byte duty;
                if (!float.TryParse(parts[2 + i * 2], NumberStyles.Float, CultureInfo.InvariantCulture, out temp))
                    return false;
                if (!byte.TryParse(parts[3 + i * 2], NumberStyles.Integer, CultureInfo.InvariantCulture, out duty))
                    return false;
                points[i] = new FanCurvePoint
                {
                    Temperature = temp,
                    DutyPercent = duty
                };
            }
            return true;
        }

        public static bool TryParseAck(string frame)
        {
            return IdentifyFrame(frame) == FrameType.Ack;
        }

        public static bool TryParseNack(string frame, out int errorCode)
        {
            errorCode = 0;
            if (IdentifyFrame(frame) != FrameType.Nack)
                return false;

            int starIdx = frame.IndexOf(ChecksumSep);
            string dataPart = frame.Substring(1, starIdx - 1);
            // NAK,CC
            string[] parts = dataPart.Split(',');
            if (parts.Length != 2) return false;
            return int.TryParse(parts[1], NumberStyles.Integer, CultureInfo.InvariantCulture, out errorCode);
        }
    }
}
